src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -917,8 +902,228 @@ impl<T: Timestamp> CollectionState<T> {
+    uppers: HashMap<GlobalId, MutableAntichain<mz_repr::Timestamp>>,
+}
+
+#[async_trait]
+impl GenericClient<StorageCommand<mz_repr::Timestamp>, StorageResponse<mz_repr::Timestamp>>
+    for StorageCommandReconcile
+{
+    async fn send(&mut self, cmd: StorageCommand<mz_repr::Timestamp>) -> Result<(), anyhow::Error> {
+        match &cmd {
+            StorageCommand::CreateSources(sources) => self
+                .sources
+                .extend(sources.iter().cloned().map(|i| (i.id, i))),
+            _ => {}
+        }
+        <_ as GenericClient<StorageCommand, StorageResponse>>::send(&mut self.client, cmd).await
+    }
+    async fn recv(&mut self) -> Result<Option<StorageResponse<mz_repr::Timestamp>>, anyhow::Error> {
+        loop {
+            match <_ as GenericClient<StorageCommand, StorageResponse>>::recv(&mut self.client)
+                .await
+            {
+                Ok(response) => break Ok(response),
+                Err(e) => {
+                    let _ = self.client.reconnect().await;
+                    <_ as GenericClient<StorageCommand, StorageResponse>>::send(
+                        &mut self.client,
+                        StorageCommand::CreateSources(self.sources.values().cloned().collect()),
+                    )
+                    .await?;
+                }
+            }
+        }
+    }
+}
+*/
+
+use crate::client::Reconnect;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio_stream::Stream;
+
+/// `ReconciledStorageClient` is the interface to "storage reconciliation", which
+/// continuously reconnects and rehydrates `storaged` instances to allow source
+/// ingestion to continue to work
+///
+/// It presents 2 primary API's: `send` allows you send `StorageCommand`s over the
+/// to storaged instance, and `recv_stream` offers a `Stream` api to receive
+/// responses back
+#[derive(Debug)]
+pub struct ReconciledStorageClient {
+    command_sender: Sender<StorageCommand>,
+    response_receiver: tokio_stream::wrappers::ReceiverStream<StorageResponse>,
+}
+
+impl ReconciledStorageClient {
+    /// Send a `StorageCommand` to the underlying client.
+    pub async fn send(&self, cmd: StorageCommand) -> Result<(), anyhow::Error> {
+        Ok(self
+            .command_sender
+            .send(cmd)
+            .await
+            .map_err(|_| anyhow::anyhow!("client task is dead"))?)
+    }
+
+    /// Receive a `StorageReponse` from the underlying client.
+    pub fn response_stream(&mut self) -> impl Stream<Item = StorageResponse> + '_ {
+        &mut self.response_receiver
+    }
+
+    /// Create a `ReconciledStorageClient` from a `StoragedRemoteClient`.
+    ///
+    /// This api may seem curious, as it takes a concrete type instead of
+    /// any `GenericClient`. This is done because `StoragedRemoteClient` is
+    /// guaranteed to present a cancel-safe API, but general `GenericClient`s
+    /// are note.
```

Comment:
As of https://github.com/MaterializeInc/materialize/pull/13441, `GenericClient::recv` implementations are required to be cancellation safe, so we can accept any client here now!

Reply:
I am hoping we still switch this to tasks+channels in the future, but that will cleanup https://github.com/MaterializeInc/materialize/pull/13360#discussion_r912420752